### ISSUE TITLE
@Value.Default behavior change with immutable collections (2.2-beta5)

### ISSUE THREAD

**Issue by nezda** [immutables/immutables#326]
In 2.1.12, if `caseSensitive()` returned `true`, below simplified code would return that `ImmutableSortedSet` with custom comparator, but in 2.2-beta5 it always returns a new `ImmutableSet` (case sensitive hash set).  My work around is to change return type to `ImmutableSortedSet` (thus can't use hash set even for case sensitive situations).

```java
  @Value.Default
  public ImmutableSet<String> words() {
    final Set<String> words;
    if (caseInsensitive()) {
      words = Sets.newTreeSet(String.CASE_INSENSITIVE_ORDER);
    } else {
      words = Sets.newHashSet();
    }
    words.addAll(Arrays.asList(
        "Foo",
        "foo",
        "BAR",
        "bar"
    ));
    if (caseInsensitive()) {
      return ImmutableSortedSet.orderedBy(String.CASE_INSENSITIVE_ORDER).addAll(words).build();
    } else {
      return ImmutableSet.copyOf(words);
    }
  }
```

**Comment by elucash** [comment]
Thank you for bringing this up! The changes were that we now allow collections to have default values, in the sense that previously defaulted collection was always turned into regular attribute with no `add`, `addAll` builder methods etc.

In order to sanitize the value returned from initializer method, we call `ImmutableSet.copyOf`. The funny things is that Guava folks are checking for this specifically

```java
...
    if (elements instanceof ImmutableSet
        && !(elements instanceof ImmutableSortedSet)) {
... 
```

The logic there is that on `ImmutableSet.copyOf(col)` the `col` object reference will be returned intact if it is `ImmutableSet` and not `ImmutableSortedSet` and not lazy view, otherwise it will be copied.
The seeming consensus is that each collection have very specific properties and one cannot expect that `ImmutableSortedSet` can be used interchangeably with `ImmutableSet` (which might be more precisely called `ImmutableLinkedHashSet`, soft of..) even if they both are `Set` and, which is even more strange, `ImmutableSortedSet extends ImmutableSet`.

On one hand we can fix this by checking it in DIY manner (of skipping the check if it is known to be immutable). On the other hand I think that the example you provided can benefit of some design changes. Ultimately I would abstract the set into special wrapper container which constructor might be used to make decisions about backing data structure. The simpler workaround may be to use `@Value.Derived` where we don't mess with the returned collection.

I'm leaning toward adding only a small change which seems have no complicating consequences, skip `copyOf` if the attribute type is declared to be immutable, but still do `copyOf` if `@Default Set<T> attribute() {..`.

**Comment by nezda** [comment]
Excellent explanation!  I like both your work arounds:

Just use `@Value.Derived` so it'll be left alone and still be cached.

and sounds like

> I'm leaning toward adding only a small change which seems have no complicating consequences, skip copyOf if the attribute type is declared to be immutable, but still do copyOf if @Default Set<T> attribute() {...

would mean my code could stay the same since it is declared `ImmutableSet` so you'd skip the `copyOf`, which sounds even better.

**Comment by elucash** [comment]
released in 2.2 Thank you!
